### PR TITLE
*: fix a data race.

### PR DESCRIPTION
**Pull request by disksing**
Operators may be accessed concurrently, should be mutex guarded.

### CODE REVIEW

File: server/operator.go [hunk 1]
```diff
@@ -154,28 +154,33 @@ type Operator interface {
...
 
 type adminOperator struct {
-	Name   string        `json:"name"`
-	Region *RegionInfo   `json:"region"`
-	Start  time.Time     `json:"start"`
-	Ops    []Operator    `json:"ops"`
-	State  OperatorState `json:"state"`
+	Name         string    `json:"name"`
+	Start        time.Time `json:"start"`
+	sync.RWMutex `json:"-"`
```
Comment (siddontang) [inline server/operator.go]: what does this mutex guarantee for?

**Comment by huachaohuang** [comment]
What cause the data race?
Seems not a good idea to add lock to every operator?

File: server/operator.go (hunk 1 above)
Comment (disksing) [inline server/operator.go]: Fields below.

**Comment by disksing** [comment]
@huachaohuang 
Here for instance: https://github.com/pingcap/pd/blob/9d2d2ae5c69a4b54ab4e4ad2cb40db11fb9c92dd/server/coordinator.go#L96-L106 
Although _getOperator_ is protected inside, we need to update `Operator` after, without any lock. If another goroutine wants to read it, then there will be a data race.

File: server/operator.go (hunk 1 above)
Comment (siddontang) [inline server/operator.go]: IMO, if you use an anonymous mutex, this mutex guarantees the whole structure.

File: server/operator.go [hunk 2]
```diff
@@ -406,7 +437,9 @@ func newTransferLeaderOperator(regionID uint64, oldLeader, newLeader *metapb.Pee
 }
 
 func (op *transferLeaderOperator) String() string {
-	return fmt.Sprintf("%+v", *op)
+	op.RLock()
+	defer op.RUnlock()
+	return jsonString(op)
```
Comment (choleraehyq) [inline server/operator.go]: These `defer`s can be refactor to:
```
op.RLock()
res := jsonString(op)
op.RUnlock()
return res
```
Cost of `defer` in such small functions may be relatively big.

Comment (disksing) [inline server/operator.go]: operators would not be performance bottleneck, so I prefer readability to performance here.

File: server/operator.go (hunk 1 above)
Comment (disksing) [inline server/operator.go]: It depends on the way you see it. It does protect the whole structure, only the readonly fields do not need to protect.

Comment (siddontang) [inline server/operator.go]: so you should put the mutex in the beginning of the struct.

Comment (huachaohuang) [inline server/operator.go]: I agree with @siddontang , it's more clear to put the mutex in the beginning to protect the whole structure.

File: server/util.go [hunk 3]
```diff
@@ -312,3 +313,8 @@ func parseTimestamp(data []byte) (time.Time, error) {
 func subTimeByWallClock(after time.Time, before time.Time) time.Duration {
 	return time.Duration(after.UnixNano() - before.UnixNano())
 }
+
+func jsonString(v interface{}) string {
+	data, _ := json.Marshal(v)
```
Comment (siddontang) [inline server/util.go]: if error, should panic?

Comment (disksing) [inline server/util.go]: Only for logging, ok to ignore it, I guess.

**Pull request merged**